File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();

        vector<int> ans;
        vector<int> prefix(n);
        vector<int> suffix(n);
        int p =1;
        int s=1;
        for(int i=0;i<n;i++){
            p*=nums[i];
            prefix[i]=p;
        }

        for(int i=n-1;i>=0;i--){
            s*=nums[i];
            suffix[i]=s;
        }
        for(int i=0;i<n;i++){
            if(i==0) ans.push_back(suffix[i+1]);
            else if(i==n-1) ans.push_back(prefix[i-1]);
            else ans.push_back(prefix[i-1]*suffix[i+1]);
        }
        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] prefix = new int[nums.length];
        int[] suffix = new int[nums.length];

        int[] ans = new int[nums.length];

        int prod1=1;
        int prod2=1;

        for(int i=0;i < nums.length; i++){
            prod1 *= nums[i];
            prefix[i] = prod1;

            prod2 *= nums[nums.length-1-i];
            suffix[nums.length-1-i] = prod2;

        }

        for(int i=0;i<nums.length;i++){
            if(i==0) ans[i] = suffix[i+1];
            else if(i==nums.length-1) ans[i] = prefix[i-1];
            else{
                ans[i] = prefix[i-1]*suffix[i+1];
            }
        }

        return ans;
    }
}

// 1, 2, 3, 4
// 1, 2, 6, 12 // prefix
// 24, 24, 12, 4 //suffix







